source/app/controllers/urls_controller.rb

```diff
@@ -1,2 +1,53 @@
 class UrlsController < ApplicationController
+  before_action :set_url, only: [:show, :edit, :update, :destroy]
+
+  def index
+  end
+
+  def show
+  end
+
+  def create
+    # validate that the user entered a url
+    if url_params[:address].size <= 3 || url_params[:address].to_s.match(/^.+\..+/).blank?
```

Comment:
I'd push all of this validation logic into the model if possible or wrap it into some sort of validator class that raises errors when params don't meet a condition then you can use `rescue_from` to handle the exceptions/render error messaging.